sql/core/src/main/scala/org/apache/spark/sql/execution/local/NestedLoopJoinNode.scala

```diff
@@ -0,0 +1,158 @@
+
+case class NestedLoopJoinNode(
+    conf: SQLConf,
+    left: LocalNode,
+    right: LocalNode,
+    buildSide: BuildSide,
+    joinType: JoinType,
+    condition: Option[Expression]) extends BinaryLocalNode(conf) {
+
+  override def output: Seq[Attribute] = {
+    joinType match {
+      case LeftOuter =>
+        left.output ++ right.output.map(_.withNullability(true))
+      case RightOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output
+      case FullOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output.map(_.withNullability(true))
+      case x =>
+        throw new IllegalArgumentException(
+          s"NestedLoopJoin should not take $x as the JoinType")
+    }
+  }
+
+  private[this] def genResultProjection: InternalRow => InternalRow = {
+    if (outputsUnsafeRows) {
+      UnsafeProjection.create(schema)
+    } else {
+      identity[InternalRow]
+    }
+  }
+
+  private[this] var currentRow: InternalRow = _
+
+  private[this] var iterator: Iterator[InternalRow] = _
+
+  override def open(): Unit = {
+    val (streamed, build) = buildSide match {
+      case BuildRight => (left, right)
+      case BuildLeft => (right, left)
+    }
+    build.open()
+    val buildRelation = new CompactBuffer[InternalRow]
+    while (build.next()) {
+      buildRelation += build.fetch().copy()
+    }
+    build.close()
+
+    val boundCondition =
+      newPredicate(condition.getOrElse(Literal(true)), left.output ++ right.output)
+
+    val leftNulls = new GenericMutableRow(left.output.size)
+    val rightNulls = new GenericMutableRow(right.output.size)
+    val joinedRow = new JoinedRow
+    val includedBuildTuples = new BitSet(buildRelation.size)
+    val resultProj = genResultProjection
+    streamed.open()
+
+    val matchesOrStreamedRowsWithNulls = streamed.asIterator.flatMap { streamedRow =>
+      val matchedRows = new CompactBuffer[InternalRow]
+
+      var i = 0
+      var streamRowMatched = false
+
+      while (i < buildRelation.size) {
+        val buildRow = buildRelation(i)
+        buildSide match {
+          case BuildRight if boundCondition(joinedRow(streamedRow, buildRow)) =>
+            matchedRows += resultProj(joinedRow(streamedRow, buildRow)).copy()
+            streamRowMatched = true
+            includedBuildTuples.set(i)
+          case BuildLeft if boundCondition(joinedRow(buildRow, streamedRow)) =>
+            matchedRows += resultProj(joinedRow(buildRow, streamedRow)).copy()
+            streamRowMatched = true
+            includedBuildTuples.set(i)
+          case _ =>
+        }
+        i += 1
+      }
+
+      (streamRowMatched, joinType, buildSide) match {
+        case (false, LeftOuter | FullOuter, BuildRight) =>
+          matchedRows += resultProj(joinedRow(streamedRow, rightNulls)).copy()
+        case (false, RightOuter | FullOuter, BuildLeft) =>
+          matchedRows += resultProj(joinedRow(leftNulls, streamedRow)).copy()
+        case _ =>
+      }
```

Comment:
it would be clearer if this looks like:
```
// If this row had no matches and we're using outer join, join it with the null row
if (!streamRowMatched) {
  (joinType, buildSide) match {
    ...
  }
}
```